src/view/dataElements/xdmObject.jsx

```diff
@@ -57,101 +69,210 @@ const XdmObject = ({
 
   // TODO: break apart this extension view
   return (
     <ExtensionView
       getInitialValues={({ initInfo }) => {
-        return fetchSchemasMeta({
+        return fetchSandboxes({
           orgId: initInfo.company.orgId,
           imsAccess: initInfo.tokens.imsAccess
         })
-          .then(_schemasMeta => {
-            setSchemasMeta(_schemasMeta);
+          .then(_sandboxesMeta => {
+            let sandboxName = null;
+            // if sandboxes are returned from the api call
+            if (_sandboxesMeta.sandboxes && _sandboxesMeta.sandboxes.length) {
+              setSandboxesMeta(_sandboxesMeta);
+              // default to the first item in the list
+              let sandboxMeta = _sandboxesMeta.sandboxes[0];
+              // if a sandbox exists in settings and is in the list, choose that item
+              if (initInfo.settings && initInfo.settings.sandbox) {
+                sandboxMeta = _sandboxesMeta.sandboxes.find(
+                  _sandboxMeta =>
+                    _sandboxMeta.name === initInfo.settings.sandbox.name
+                );
+              }
+              if (sandboxMeta && sandboxMeta.name) {
+                sandboxName = sandboxMeta.name;
+              }
+              setSelectedSandboxMeta(sandboxMeta);
+            }
+
+            return fetchSchemasMeta({
+              orgId: initInfo.company.orgId,
+              imsAccess: initInfo.tokens.imsAccess,
+              sandboxName
+            });
+          })
+          .then(response => {
+            setSchemasMeta(response.results);
 
-            if (!_schemasMeta.length) {
+            if (!response.results.length) {
+              setSchemasMetaStatus(STATUS_ERROR || null);
               return {};
             }
 
+            setSchemasMetaStatus(STATUS_LOADED || null);
+
             // TODO: don't initialize form state if the schema doesn't match
-            let schemaMeta = _schemasMeta[0];
+            let schemaMeta = response.results[0];
 
+            // if a schema exists in settings, use it
             if (initInfo.settings && initInfo.settings.schema) {
-              schemaMeta = _schemasMeta.find(
+              schemaMeta = response.results.find(
                 _schemaMeta => _schemaMeta.$id === initInfo.settings.schema.id
               );
             }
 
             setSelectedSchemaMeta(schemaMeta);
+            setSchemaStatus(STATUS_LOADING || null);
 
+            const { sandboxName } = response;
             return fetchSchema({
               orgId: initInfo.company.orgId,
               imsAccess: initInfo.tokens.imsAccess,
-              schemaMeta
-            });
+              schemaMeta,
+              sandboxName
+            })
+              .then(schema => {
+                setSchemaStatus(STATUS_LOADED || null);
+
+                return schema;
+              })
+              .catch(() => {
+                setSchemaStatus(STATUS_ERROR || null);
+              });
```

Comment:
It looks like you can pull the `.then` and `.catch` up to the top-level of the promise chain rather than having it nested.

Reply:
If you do move the `.catch` up, just be aware that it will catch errors from `fetchSchema` but also any errors that happened higher up the chain, like errors from `fetchSchemasMeta`. You'll need to think through how you want to deal with the various errors that may occur anywhere in this promise chain.

Reply:
I have the null to get around this error that WebStorm is telling me without `|| null`:

```Argument type string is not assignable to parameter type ((prevState: undefined) => undefined) | undefined   Type string is not assignable to type (prevState: undefined) => undefined ```.  Can you suggest an alternative?

Reply:
Roger